File: packages/provide-uterm-go/controlplane/memory/graphical_target_store.go
// SPDX-License-Identifier: AGPL-3.0-or-later

package memory

import (
	"context"
	"sort"

	cp "github.com/provide-io/provide-uterm/packages/provide-uterm-go/controlplane"
)

type graphicalTargetStore struct{ state *State }

func (s *graphicalTargetStore) Put(_ context.Context, rec cp.GraphicalTargetRecord) error {
	s.state.GraphicalTargets[rec.TargetID] = rec
	return nil
}

func (s *graphicalTargetStore) Get(_ context.Context, targetID string) (*cp.GraphicalTargetRecord, error) {
	rec, ok := s.state.GraphicalTargets[targetID]
	if !ok {
		return nil, nil
	}
	return &rec, nil
}

func (s *graphicalTargetStore) List(context.Context) ([]cp.GraphicalTargetRecord, error) {
	out := make([]cp.GraphicalTargetRecord, 0, len(s.state.GraphicalTargets))
	for _, rec := range s.state.GraphicalTargets {
		out = append(out, rec)
	}
	sort.Slice(out, func(i, j int) bool { return out[i].TargetID < out[j].TargetID })
	return out, nil
}

func (s *graphicalTargetStore) Delete(_ context.Context, targetID string) (bool, error) {
	_, ok := s.state.GraphicalTargets[targetID]
	delete(s.state.GraphicalTargets, targetID)
	return ok, nil
}

File: packages/provide-uterm/src/provide/uterm/control/plane/memory/graphical_target_store.py
#
# SPDX-FileCopyrightText: Copyright (c) 2025-2026 provide.io llc. All rights reserved.
# SPDX-License-Identifier: AGPL-3.0-or-later
#
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from provide.uterm.control.plane.graphical_target.types import GraphicalTargetRecord
    from provide.uterm.control.plane.memory.transaction import MemoryState, MemoryTransaction


class MemoryGraphicalTargetStore:
    def __init__(self, state: MemoryState, tx: MemoryTransaction) -> None:
        self._state = state
        self._tx = tx

    async def put_graphical_target(self, record: GraphicalTargetRecord) -> None:
        self._state.graphical_targets[record.target_id] = record

    async def get_graphical_target(self, target_id: str) -> GraphicalTargetRecord | None:
        return self._state.graphical_targets.get(target_id)

    async def list_graphical_targets(self) -> list[GraphicalTargetRecord]:
        # Sorted by target_id so the memory and SQLite backends agree on order;
        # the SQLite store gets it from ORDER BY, dict order is insertion order.
        return [self._state.graphical_targets[key] for key in sorted(self._state.graphical_targets)]

    async def delete_graphical_target(self, target_id: str) -> bool:
        return self._state.graphical_targets.pop(target_id, None) is not None
